Source/Scene/LabelCollection.js

```diff
@@ -215,46 +215,91 @@ define([
     var glyphPixelOffset = new Cartesian2();
 
     function repositionAllGlyphs(label, resolutionScale) {
+        
         var glyphs = label._glyphs;
+        var text = label._text;
+
         var glyph;
         var dimensions;
-        var totalWidth = 0;
-        var maxHeight = 0;
+
+        var lineWidth = 0;
+        var maxLineWidth = 0;
+
+        var maxGlyphHeight = 0;
+        var maxGlyphDescent = 0;
+
+        var numberOfLines = 1;
 
         var glyphIndex = 0;
         var glyphLength = glyphs.length;
         for (glyphIndex = 0; glyphIndex < glyphLength; ++glyphIndex) {
             glyph = glyphs[glyphIndex];
             dimensions = glyph.dimensions;
-            totalWidth += dimensions.computedWidth;
-            maxHeight = Math.max(maxHeight, dimensions.height);
+            
+            lineWidth += dimensions.width;
+            maxLineWidth = Math.max(maxLineWidth, lineWidth);
+            maxGlyphHeight = Math.max(maxGlyphHeight, dimensions.height);
+            maxGlyphDescent = Math.max(maxGlyphDescent, dimensions.descent);
+
+            if (text.charAt(glyphIndex) === '\n') {
+                numberOfLines += 1;
+                lineWidth = 0;
+            }
         }
 
         var scale = label._scale;
         var horizontalOrigin = label._horizontalOrigin;
         var widthOffset = 0;
         if (horizontalOrigin === HorizontalOrigin.CENTER) {
-            widthOffset -= totalWidth / 2 * scale;
+            widthOffset -= maxLineWidth / 2 * scale;
         } else if (horizontalOrigin === HorizontalOrigin.RIGHT) {
-            widthOffset -= totalWidth * scale;
+            widthOffset -= maxLineWidth * scale;
+        }
+
+        var heightOffset = 0;
+
+
+        var totalHeight = (maxGlyphHeight * numberOfLines) + (maxGlyphDescent * (numberOfLines-1));
+
+        var heightReference = label._heightReference;
+        var verticalOrigin = (heightReference === HeightReference.NONE) ? label._verticalOrigin : VerticalOrigin.BOTTOM;
+        if (verticalOrigin === VerticalOrigin.CENTER) {
+            // Subtract maxGlyphDescent for backwards compatibility
+            heightOffset += (totalHeight / 2 * scale) - maxGlyphDescent;
+        }
+        else if (verticalOrigin === VerticalOrigin.BOTTOM) {
+            // Subtract maxGlyphHeight for backwards compatibility
+            heightOffset += (totalHeight * scale) - maxGlyphHeight;
+        }
+        else if (verticalOrigin === VerticalOrigin.TOP) {
```

Comment:
Also, update the formatting in the block above this while you're in there.  It should be
```
} else if () {
```